Module8

ViktoriiaLozova opened the pull request.

victorialozova committed:
/* Module 8
Задание 1
Осуществить сжатие английского текста, заменив каждую группу из двух или более рядом стоящих символов,
на один символ, за которым следует количество его вхождений в группу. К примеру, строка hellowoooorld должна сжиматься в hel2owo4rld.
В случае если в тексте будет присутствовать цифра в единичном экземпляре перед ней необходимо поставить цифру 1.
Например, строка He4lo будет преобразована в He14lo

Также если количество повторений будет превышать 9 используйте несколько повторений.
Например, строка heloooooooooo будет преобразована в hel9oo. Строка heloooooоooooo будет преобразована в  hel9o2o

Реализуйте возможность распаковки текста, сжатого по этому правилу. В случае если распаковка не возможна метод распаковки
должен выбросить checked exception.
 */

victorialozova committed:
Module 8
/* Задание 1
Осуществить сжатие английского текста, заменив каждую группу из двух или более рядом стоящих символов,
на один символ, за которым следует количество его вхождений в группу. К примеру, строка hellowoooorld должна сжиматься в hel2owo4rld.
В случае если в тексте будет присутствовать цифра в единичном экземпляре перед ней необходимо поставить цифру 1.
Например, строка He4lo будет преобразована в He14lo

Также если количество повторений будет превышать 9 используйте несколько повторений.
Например, строка heloooooooooo будет преобразована в hel9oo. Строка heloooooоooooo будет преобразована в  hel9o2o

Реализуйте возможность распаковки текста, сжатого по этому правилу. В случае если распаковка не возможна метод распаковки
должен выбросить checked exception.
 */

victorialozova committed:
Module 8
/*Задание 3
Используйте классы задания 2 модуля 3 для реализации задачи поиска человека по паспортным данным.
Предусмотрите новый класс Passport в котором будут серия и номер паспорта (серия типа String, номер типа Integer).
Создайте несколько экземпляров Passport и Person для установления связи между ними используйте HashMap<Passport, Person>
Для поиска создайте новый экземпляр Passport в который внесите заведомо существующие данные (совпадающие с данными
Passport на предыдущем шаге), проверьте поведение приложения если паспортные данные внесены с ошибкой.
Обратите внимание на поведение HashMap в случае если Passport не имеет методов hashCode и equals
*/

AntonApanasovych commented on src/main/java/module8/Passport.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Incorrect implementation for `equals`. You should also check for `null` and self reference.
I recommend to use IDEA equals and hashcode generator. Try combination `alt + insert`, displayed context menu should contain equals and hashcode generator.

ViktoriiaLozova replied:
Done. Hashcode and equals generator was used

AntonApanasovych commented on src/main/java/module8/Passport.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Override
+    public int hashCode() {
```
hashcode is correct but it is good to use generator for both `equals` and `hashcode`.

AntonApanasovych commented on src/main/java/module8/Task1.java:
```diff
@@ -0,0 +1,90 @@
+            if (charFromstring == '1' && !isNumber(source.charAt(i + 1)) && source.charAt(i - 1) != '1'
+                    || isNumber(charFromstring) && charFromstring != '1' && i == 0
+                    || charFromstring == '0' && source.charAt(i - 1) != '1') {
+                throw new Exception("Decoding error");
```
Do not use `Exception` for runtime errors. It is checked exception and compiler will force you to handle it in any method that will call your method. And also it is too broad.
More compliant solution to use `RuntimeException`. But it is also too broad, better to use concrete or even custom exception, i.e. IllegalArgumentException, IllegalStateException etc. If you will not be able to find concrete exception use `RuntimeException`

ViktoriiaLozova replied:
Done. new Exception was changed to DecodingException

AntonApanasovych commented on src/main/java/module8/Task3.java:
```diff
@@ -0,0 +1,35 @@
+
+public class Task3 {
+    public static void main(String[] args) {
+        HashMap<Passport, Person> map = new HashMap<>();
```
`Map<Passport, Person> map = new HashMap<>();`

ViktoriiaLozova replied:
Done. HashMap was changed to Map

victorialozova committed:
Module 8
/*
- Equals and hashcode generator was used
- Hashmap was changed to Map
- new Exception was changed to DecodingException
*/
